src/coreclr/jit/lower.cpp

```diff
@@ -1842,6 +1842,155 @@ GenTree* Lowering::AddrGen(void* addr)
+        assert(sig.sigInst.methInstCount == 1);
+        lengthScale = genTypeSize(
+            JITtype2varType(comp->info.compCompHnd->getTypeForPrimitiveValueClass(sig.sigInst.methInst[0])));
+
+        // If value is not zero, we can only unroll for single-byte values
+        if (!valueArg->IsIntegralConst(0) && (lengthScale != 1))
+        {
+            JITDUMP("SpanHelpers.Fill's value is not unroll-friendly - bail out.\n")
+            return false;
+        }
+    }
+    else
+    {
+        // void SpanHelpers::ClearWithoutReferences(ref byte b, nuint byteLength)
+        //
+        assert(call->gtArgs.CountUserArgs() == 2);
+        assert(ni == NI_System_SpanHelpers_ClearWithoutReferences);
+
+        lengthScale = 1; // it's always in bytes
+    }
+
+    // Convert lenCns to bytes
+    ssize_t lenCns = lengthArg->AsIntCon()->IconValue();
+    if (CheckedOps::MulOverflows((target_ssize_t)lenCns, (target_ssize_t)lengthScale, CheckedOps::Signed))
+    {
+        // lenCns overflows
+        JITDUMP("lenCns * lengthScale overflows - bail out.\n")
+        return false;
+    }
+    lenCns *= (ssize_t)lengthScale;
+
+    // TODO-CQ: drop the whole thing in case of lenCns = 0
+    if ((lenCns <= 0) || (lenCns > (ssize_t)comp->getUnrollThreshold(Compiler::UnrollKind::Memset)))
+    {
+        JITDUMP("Size is either 0 or too big to unroll - bail out.\n")
+        return false;
+    }
+
+    JITDUMP("Accepted for unrolling!\nOld tree:\n");
+    DISPTREE(call);
+
+    GenTree* blkInnerValue = nullptr;
+    GenTree* blkValue;
+    if (valueArg == nullptr || valueArg->IsIntegralConst(0))
+    {
+        // Simple zeroing
+        blkValue = comp->gtNewZeroConNode(TYP_INT);
+        blkValue->SetContained();
+    }
+    else
+    {
+        // Non-zero (byte) value
+        blkInnerValue = comp->gtCloneExpr(valueArg);
+        blkInnerValue->SetContained();
+        blkValue = comp->gtNewOperNode(GT_INIT_VAL, TYP_INT, blkInnerValue);
+        blkValue->SetContained();
+    }
+
+    GenTreeBlk* storeBlk =
+        comp->gtNewStoreBlkNode(comp->typGetBlkLayout((unsigned)lenCns), dstRefArg, blkValue, GTF_IND_UNALIGNED);
+    storeBlk->gtBlkOpKind = GenTreeBlk::BlkOpKindUnroll;
+
+    // Insert/Remove trees into LIR
+    BlockRange().InsertBefore(call, blkValue);
+    BlockRange().InsertBefore(call, storeBlk);
+    BlockRange().Remove(lengthArg);
+    BlockRange().Remove(call);
+    if (valueArg != nullptr)
+    {
+        // valueArg is just a constant in case of Fill<T>
+        // and doesn't exist in case of ClearWithoutReferences
+        assert(valueArg->IsIntegralConst());
+        BlockRange().Remove(valueArg);
+    }
+    if (blkInnerValue != nullptr)
+    {
+        BlockRange().InsertBefore(blkValue, blkInnerValue);
+    }
+
+    // Remove all non-user args (e.g. r2r cell)
+    for (CallArg& arg : call->gtArgs.Args())
+    {
+        if (arg.IsArgAddedLate())
+        {
+            arg.GetNode()->SetUnusedValue();
+        }
+    }
```

Comment:
Instead of this (which is a bit fragile, and `IsArgAddedLate` is not the right thing to check here), can you change the above to `BlockRange().Remove(call, true)` and then `ClearUnusedValue()` on the nodes still used?

Reply:
That doesn't sound right? `BlockRange().Remove(call, true)` will recursively mark all subtrees in our args as unused, won't it? Do we want to then recursively mark them as used again (can some of them be unused before the Remove?)

Reply:
> `BlockRange().Remove(call, true)` will recursively mark all subtrees in our args as unused, won't it?

I don't think so.

Reply:
> > `BlockRange().Remove(call, true)` will recursively mark all subtrees in our args as unused, won't it?
> 
> I don't think so.

![image](https://github.com/dotnet/runtime/assets/523221/3c2c2712-7777-4a43-9a98-b06b288913a4)

it uses a visitor, doesn't?

Reply:
`VisitOperands` isn't recursive

Reply:
Got it, thanks!